File: src/ipc/messages/types/structs/PathAggregate.ts
import * as ldm from '../../leapdatamodel';
import * as valuetype from '../../valuetype';
import type * as types from '..';
import { StructBase as _StructBase } from '../../StructBase';
import { LeapError as _LeapError } from '../../LeapError';
import { ValueBase as _ValueBase } from '../../ValueBase';

export class PathAggregate extends _StructBase {
    static _type: string = 'path-aggregate';
    static _leapStruct: valuetype.LeapStruct = new valuetype.LeapStruct("path-aggregate", [], [new valuetype.Property("entries", new valuetype.ValueType("list", [new valuetype.ValueType("entry", [])])), new valuetype.Property("tree", new valuetype.ValueType("list", [new valuetype.ValueType("list", [new valuetype.ValueType("int", [])])]))]);
    entries: Array<types.structs.Entry>;
    tree: Array<Array<number>>;

    constructor(entries: Array<types.structs.Entry>, tree: Array<Array<number>>, ) {
        super();
        this.entries = entries;
        this.tree = tree;
    }

    _verifyAsTypeOrRaise(path: string[], valueType: valuetype.ValueType): void {
        if (valueType.name == PathAggregate._type) {
            const structType = PathAggregate._leapStruct.applyArgs(valueType.args);
            _ValueBase._verifyValueOrRaise(path.concat(['entries']), this.entries, structType.props[0].propType);
            _ValueBase._verifyValueOrRaise(path.concat(['tree']), this.tree, structType.props[1].propType);
        } else {
            throw _ValueBase._pathError(path, `expecting ${PathAggregate._type}`);
        }
    }

    _toLdm(valueType: valuetype.ValueType): ldm.Value {
        let value = new Map<string, ldm.Value>();
        const structType = PathAggregate._leapStruct.applyArgs(valueType.args);
        value.set("entries", _ValueBase._valueToLdm(this.entries, structType.props[0].propType));
        value.set("tree", _ValueBase._valueToLdm(this.tree, structType.props[1].propType));
        return new ldm.StructValue(value, valueType);
    }
}

File: src-tauri/src/ipc/messages/types/structs/path_aggregate.rs
use std::collections::HashMap as _HashMap;
#[allow(unused_imports)]
use super::super::super::types;
use super::super::super::value_base::ValueBase as _ValueBase;
use super::super::super::struct_base::StructBase as _StructBase;
use super::super::super::leap_data_model as ldm;
use super::super::super::value_type::ValueType as _ValueType;

#[derive(Clone, Debug)]
pub struct PathAggregate {
    pub entries: Vec<types::structs::Entry>,
    pub tree: Vec<Vec<i64>>,
}

impl PathAggregate {
    pub fn new(entries: Vec<types::structs::Entry>, tree: Vec<Vec<i64>>, ) -> Self {
        Self {
            entries,
            tree,
        }
    }
}

impl _ValueBase for PathAggregate {
    const _TYPE: &'static str = "path-aggregate";

    fn _applied_args() -> Vec<_ValueType> {
        vec![
        ]
    }

    fn _to_ldm(&self) -> ldm::Value {
        ldm::Value::Struct(self._to_ldm_struct())
    }
}

impl _StructBase for PathAggregate {
    fn _to_ldm_struct(&self) -> ldm::StructValue {
        let mut value = _HashMap::new();
        value.insert("entries".to_owned(), self.entries._to_ldm());
        value.insert("tree".to_owned(), self.tree._to_ldm());
        ldm::StructValue{
            value,
            value_type: _ValueType::new(Self::_TYPE.to_owned(), Self::_applied_args()),
        }
    }
}